Panic when using `example` or `enum` struct tag for OmittableNullable

I'm using the support for optional body fields as implemented in #145 

The OpenAPI spec is generated properly when using a definition like:
`Hash OmittableNullable[string] json:"hash,omitempty" doc:"Hash of the file contents of the artifact, when available"`

After adding the `example` struct tag like `example:"47944637b32facc94c1c49b32202b63e23b6ea7f225baa1d89868b3ff32d4a1e"`, a panic is generated. Similarly, a panic occurs when using an `enum` struct tag like `enum:"MD5,SHA-1,SHA-256,SHA-512"`


That same `example` struct tag works perfectly fine when the struct property is defined as `Hash string`.

```
panic: invalid tag for field 'Hash': invalid character 'b' after top-level value

goroutine 1 [running]:
github.com/danielgtaylor/huma/v2.jsonTagValue({{0x1033206fa, 0x4}, {0x0, 0x0}, {0x103386ec8, 0x103336880}, {0x103320700, 0x96}, 0x48, {0x140001a5058, ...}, ...}, ...)
	/Users/ethan.brown/go/pkg/mod/github.com/danielgtaylor/huma/v2@v2.3.0/schema.go:268 +0x828
github.com/danielgtaylor/huma/v2.jsonTag({{0x1033206fa, 0x4}, {0x0, 0x0}, {0x103386ec8, 0x103336880}, {0x103320700, 0x96}, 0x48, {0x140001a5058, ...}, ...}, ...)
	/Users/ethan.brown/go/pkg/mod/github.com/danielgtaylor/huma/v2@v2.3.0/schema.go:310 +0xcc
github.com/danielgtaylor/huma/v2.SchemaFromField({0x1033801a0?, 0x14000274540?}, {{0x1033206fa, 0x4}, {0x0, 0x0}, {0x103386ec8, 0x103336880}, {0x103320700, 0x96}, ...}, ...)
	/Users/ethan.brown/go/pkg/mod/github.com/danielgtaylor/huma/v2@v2.3.0/schema.go:351 +0x390
github.com/danielgtaylor/huma/v2.SchemaFromType({0x1033801a0, 0x14000274540}, {0x103386ec8, 0x103358720})
	/Users/ethan.brown/go/pkg/mod/github.com/danielgtaylor/huma/v2@v2.3.0/schema.go:557 +0xb00
github.com/danielgtaylor/huma/v2.(*mapRegistry).Schema(0x14000274540, {0x103386ec8, 0x103358720}, 0x1, {0x140001ec820, 0x1b})
	/Users/ethan.brown/go/pkg/mod/github.com/danielgtaylor/huma/v2@v2.3.0/registry.go:103 +0x228
github.com/danielgtaylor/huma/v2.SchemaFromField({0x1033801a0?, 0x14000274540?}, {{0x103284751, 0x4}, {0x0, 0x0}, {0x103386ec8, 0x103358720}, {0x0, 0x0}, ...}, ...)
	/Users/ethan.brown/go/pkg/mod/github.com/danielgtaylor/huma/v2@v2.3.0/schema.go:322 +0x80
github.com/danielgtaylor/huma/v2.Register[...]({_, _}, {{0x103199f7a, 0x4}, {0x1031a381d, 0x16}, 0x0, 0x0, 0x0, {0x0, ...}, ...}, ...)
	/Users/ethan.brown/go/pkg/mod/github.com/danielgtaylor/huma/v2@v2.3.0/huma.go:491 +0x2f8
github.com/train-artifact-index-service/internal/app/handler.ApiV1(0x1400029f948)
```

--- a/huma.go
+++ b/huma.go
@@ -120,7 +120,7 @@ func findParams(registry Registry, op *Operation, t reflect.Type) *findResult[*p
 
 		var example any
 		if e := f.Tag.Get("example"); e != "" {
-			example = jsonTagValue(f, f.Type, f.Tag.Get("example"))
+			example = jsonTagValue(registry, f.Type.Name(), pfi.Schema, f.Tag.Get("example"))
 		}
 
 		if def := f.Tag.Get("default"); def != "" {
@@ -190,13 +190,14 @@ func findResolvers(resolverType, t reflect.Type) *findResult[bool] {
 	}, nil)
 }
 
-func findDefaults(t reflect.Type) *findResult[any] {
+func findDefaults(registry Registry, t reflect.Type) *findResult[any] {
 	return findInType(t, nil, func(sf reflect.StructField, i []int) any {
 		if d := sf.Tag.Get("default"); d != "" {
 			if sf.Type.Kind() == reflect.Pointer {
 				panic("pointers cannot have default values")
 			}
-			return jsonTagValue(sf, sf.Type, d)
+			s := registry.Schema(sf.Type, true, "")
+			return convertType(sf.Type.Name(), sf.Type, jsonTagValue(registry, sf.Name, s, d))
 		}
 		return nil
 	})
@@ -560,7 +561,7 @@ func Register[I, O any](api API, op Operation, handler func(context.Context, *I)
 	}
 
 	resolvers := findResolvers(resolverType, inputType)
-	defaults := findDefaults(inputType)
+	defaults := findDefaults(registry, inputType)
 
 	if op.Responses == nil {
 		op.Responses = map[string]*Response{}

--- a/registry.go
+++ b/registry.go
@@ -112,6 +112,9 @@ func (r *mapRegistry) Schema(t reflect.Type, allowRef bool, hint string) *Schema
 }
 
 func (r *mapRegistry) SchemaFromRef(ref string) *Schema {
+	if !strings.HasPrefix(ref, r.prefix) {
+		return nil
+	}
 	return r.schemas[ref[len(r.prefix):]]
 }
 

--- a/schema.go
+++ b/schema.go
@@ -280,26 +280,64 @@ func floatTag(f reflect.StructField, tag string) *float64 {
 	return nil
 }
 
-func jsonTagValue(f reflect.StructField, t reflect.Type, value string) any {
-	// Special case: strings don't need quotes.
-	if t.Kind() == reflect.String || (t.Kind() == reflect.Pointer && t.Elem().Kind() == reflect.String) {
-		return value
-	}
-
-	// Special case: array of strings with comma-separated values and no quotes.
-	if t.Kind() == reflect.Slice && t.Elem().Kind() == reflect.String && value[0] != '[' {
-		values := []string{}
-		for _, s := range strings.Split(value, ",") {
-			values = append(values, strings.TrimSpace(s))
+// ensureType panics if the given value does not match the JSON Schema type.
+func ensureType(r Registry, fieldName string, s *Schema, value string, v any) {
+	if s.Ref != "" {
+		s = r.SchemaFromRef(s.Ref)
+		if s == nil {
+			// We may not have access to this type, e.g. custom schema provided
+			// by the user with remote refs. Skip validation.
+			return
 		}
-		return values
 	}
 
-	var v any
-	if err := json.Unmarshal([]byte(value), &v); err != nil {
-		panic(fmt.Errorf("invalid tag for field '%s': %w", f.Name, err))
-	}
+	switch s.Type {
+	case TypeBoolean:
+		if _, ok := v.(bool); !ok {
+			panic(fmt.Errorf("invalid boolean tag value '%s' for field '%s': %w", value, fieldName, ErrSchemaInvalid))
+		}
+	case TypeInteger, TypeNumber:
+		if _, ok := v.(float64); !ok {
+			panic(fmt.Errorf("invalid number tag value '%s' for field '%s': %w", value, fieldName, ErrSchemaInvalid))
+		}
 
+		if s.Type == TypeInteger {
+			if v.(float64) != float64(int(v.(float64))) {
+				panic(fmt.Errorf("invalid integer tag value '%s' for field '%s': %w", value, fieldName, ErrSchemaInvalid))
+			}
+		}
+	case TypeString:
+		if _, ok := v.(string); !ok {
+			panic(fmt.Errorf("invalid string tag value '%s' for field '%s': %w", value, fieldName, ErrSchemaInvalid))
+		}
+	case TypeArray:
+		if _, ok := v.([]any); !ok {
+			panic(fmt.Errorf("invalid array tag value '%s' for field '%s': %w", value, fieldName, ErrSchemaInvalid))
+		}
+
+		if s.Items != nil {
+			for i, item := range v.([]any) {
+				b, _ := json.Marshal(item)
+				ensureType(r, fieldName+"["+strconv.Itoa(i)+"]", s.Items, string(b), item)
+			}
+		}
+	case TypeObject:
+		if _, ok := v.(map[string]any); !ok {
+			panic(fmt.Errorf("invalid object tag value '%s' for field '%s': %w", value, fieldName, ErrSchemaInvalid))
+		}
+
+		for name, prop := range s.Properties {
+			if val, ok := v.(map[string]any)[name]; ok {
+				b, _ := json.Marshal(val)
+				ensureType(r, fieldName+"."+name, prop, string(b), val)
+			}
+		}
+	}
+}
+
+// convertType panics if the given value does not match or cannot be converted
+// to the field's Go type.
+func convertType(fieldName string, t reflect.Type, v any) any {
 	vv := reflect.ValueOf(v)
 	tv := reflect.TypeOf(v)
 	if v != nil && tv != t {
@@ -310,14 +348,14 @@ func jsonTagValue(f reflect.StructField, t reflect.Type, value string) any {
 			tmp := reflect.MakeSlice(t, 0, vv.Len())
 			for i := 0; i < vv.Len(); i++ {
 				if !vv.Index(i).Elem().Type().ConvertibleTo(t.Elem()) {
-					panic(fmt.Errorf("unable to convert %v to %v for field '%s': %w", vv.Index(i).Interface(), t.Elem(), f.Name, ErrSchemaInvalid))
+					panic(fmt.Errorf("unable to convert %v to %v for field '%s': %w", vv.Index(i).Interface(), t.Elem(), fieldName, ErrSchemaInvalid))
 				}
 
 				tmp = reflect.Append(tmp, vv.Index(i).Elem().Convert(t.Elem()))
 			}
 			v = tmp.Interface()
 		} else if !tv.ConvertibleTo(deref(t)) {
-			panic(fmt.Errorf("unable to convert %v to %v for field '%s': %w", tv, t, f.Name, ErrSchemaInvalid))
+			panic(fmt.Errorf("unable to convert %v to %v for field '%s': %w", tv, t, fieldName, ErrSchemaInvalid))
 		}
 
 		converted := reflect.ValueOf(v).Convert(deref(t))
@@ -330,16 +368,47 @@ func jsonTagValue(f reflect.StructField, t reflect.Type, value string) any {
 		}
 		v = converted.Interface()
 	}
+	return v
+}
+
+func jsonTagValue(r Registry, fieldName string, s *Schema, value string) any {
+	if s.Ref != "" {
+		s = r.SchemaFromRef(s.Ref)
+		if s == nil {
+			return nil
+		}
+	}
+
+	// Special case: strings don't need quotes.
+	if s.Type == TypeString {
+		return value
+	}
+
+	// Special case: array of strings with comma-separated values and no quotes.
+	if s.Type == TypeArray && s.Items != nil && s.Items.Type == TypeString && value[0] != '[' {
+		values := []string{}
+		for _, s := range strings.Split(value, ",") {
+			values = append(values, strings.TrimSpace(s))
+		}
+		return values
+	}
+
+	var v any
+	if err := json.Unmarshal([]byte(value), &v); err != nil {
+		panic(fmt.Errorf("invalid %s tag value '%s' for field '%s': %w", s.Type, value, fieldName, err))
+	}
+
+	ensureType(r, fieldName, s, value, v)
 
 	return v
 }
 
 // jsonTag returns a value of the schema's type for the given tag string.
 // Uses JSON parsing if the schema is not a string.
-func jsonTag(f reflect.StructField, name string) any {
+func jsonTag(r Registry, f reflect.StructField, s *Schema, name string) any {
 	t := f.Type
 	if value := f.Tag.Get(name); value != "" {
-		return jsonTagValue(f, t, value)
+		return convertType(f.Name, t, jsonTagValue(r, f.Name, s, value))
 	}
 	return nil
 }
@@ -378,20 +447,22 @@ func SchemaFromField(registry Registry, f reflect.StructField, hint string) *Sch
 	if enc := f.Tag.Get("encoding"); enc != "" {
 		fs.ContentEncoding = enc
 	}
-	fs.Default = jsonTag(f, "default")
+	fs.Default = jsonTag(registry, f, fs, "default")
 
-	if e := jsonTag(f, "example"); e != nil {
-		fs.Examples = []any{e}
+	if value := f.Tag.Get("example"); value != "" {
+		if e := jsonTagValue(registry, f.Name, fs, value); e != nil {
+			fs.Examples = []any{e}
+		}
 	}
 
 	if enum := f.Tag.Get("enum"); enum != "" {
-		fType := f.Type
-		if fs.Type == TypeArray {
-			fType = fType.Elem()
+		s := fs
+		if s.Type == TypeArray {
+			s = s.Items
 		}
 		enumValues := []any{}
 		for _, e := range strings.Split(enum, ",") {
-			enumValues = append(enumValues, jsonTagValue(f, fType, e))
+			enumValues = append(enumValues, jsonTagValue(registry, f.Name, s, e))
 		}
 		if fs.Type == TypeArray {
 			fs.Items.Enum = enumValues

--- a/schema_test.go
+++ b/schema_test.go
@@ -26,6 +26,26 @@ type Embedded struct {
 	Value string `json:"value" doc:"new doc"`
 }
 
+type CustomSchema struct{}
+
+func (c CustomSchema) Schema(r huma.Registry) *huma.Schema {
+	return &huma.Schema{
+		Type: "string",
+	}
+}
+
+var _ huma.SchemaProvider = CustomSchema{}
+
+type BadRefSchema struct{}
+
+func (c BadRefSchema) Schema(r huma.Registry) *huma.Schema {
+	return &huma.Schema{
+		Ref: "bad",
+	}
+}
+
+var _ huma.SchemaProvider = BadRefSchema{}
+
 func TestSchema(t *testing.T) {
 	bitSize := strconv.Itoa(bits.UintSize)
 
@@ -277,6 +297,22 @@ func TestSchema(t *testing.T) {
 			}`,
 		},
 		{
+			name: "field-default-string-pointer",
+			input: struct {
+				Value *string `json:"value,omitempty" default:"foo"`
+			}{},
+			expected: `{
+				"type": "object",
+				"properties": {
+					"value": {
+						"type": "string",
+						"default": "foo"
+					}
+				},
+				"additionalProperties": false
+			}`,
+		},
+		{
 			name: "field-default-array-string",
 			input: struct {
 				Value []string `json:"value" default:"foo,bar"`
@@ -318,6 +354,57 @@ func TestSchema(t *testing.T) {
 			}`,
 		},
 		{
+			name: "field-default-duration",
+			input: struct {
+				Value time.Duration `json:"value" default:"5000"`
+			}{},
+			expected: `{
+				"type": "object",
+				"properties": {
+					"value": {
+						"type": "integer",
+						"format": "int64",
+						"default": 5000
+					}
+				},
+				"additionalProperties": false,
+				"required": ["value"]
+			}`,
+		},
+		{
+			name: "field-example-custom",
+			input: struct {
+				Value CustomSchema `json:"value" example:"foo"`
+			}{},
+			expected: `{
+				"type": "object",
+				"properties": {
+					"value": {
+						"type": "string",
+						"examples": ["foo"]
+					}
+				},
+				"additionalProperties": false,
+				"required": ["value"]
+			}`,
+		},
+		{
+			name: "field-enum-custom",
+			input: struct {
+				Value OmittableNullable[string] `json:"value,omitempty" enum:"foo,bar"`
+			}{},
+			expected: `{
+				"type": "object",
+				"properties": {
+					"value": {
+						"type": "string",
+						"enum": ["foo", "bar"]
+					}
+				},
+				"additionalProperties": false
+			}`,
+		},
+		{
 			name: "field-any",
 			input: struct {
 				Value any `json:"value" doc:"Some value"`
@@ -331,6 +418,31 @@ func TestSchema(t *testing.T) {
 				},
 				"additionalProperties": false,
 				"required": ["value"]
+			}`,
+		},
+		{
+			// Bad ref should not panic, but should be ignored. These could be valid
+			// custom schemas that Huma won't understand.
+			name: "field-custom-bad-ref",
+			input: struct {
+				Value  BadRefSchema `json:"value" example:"true"`
+				Value2 struct {
+					Foo BadRefSchema `json:"foo"`
+				} `json:"value2" example:"{\"foo\": true}"`
+			}{},
+			expected: `{
+				"type": "object",
+				"properties": {
+					"value": {
+						"$ref": "bad"
+					},
+					"value2": {
+						"$ref": "#/components/schemas/Value2Struct",
+						"examples": [{"foo": true}]
+					}
+				},
+				"additionalProperties": false,
+				"required": ["value", "value2"]
 			}`,
 		},
 		{
@@ -420,14 +532,67 @@ func TestSchema(t *testing.T) {
 			input: struct {
 				Value int `json:"value" default:"bad"`
 			}{},
-			panics: `invalid tag for field 'Value': invalid character 'b' looking for beginning of value`,
+			panics: `invalid integer tag value 'bad' for field 'Value': invalid character 'b' looking for beginning of value`,
 		},
 		{
-			name: "panic-json-type",
+			name: "panic-json-bool",
 			input: struct {
-				Value int `json:"value" example:"true"`
+				Value bool `json:"value" default:"123"`
 			}{},
-			panics: `unable to convert bool to int for field 'Value': schema is invalid`,
+			panics: `invalid boolean tag value '123' for field 'Value': schema is invalid`,
+		},
+		{
+			name: "panic-json-int",
+			input: struct {
+				Value int `json:"value" default:"true"`
+			}{},
+			panics: `invalid number tag value 'true' for field 'Value': schema is invalid`,
+		},
+		{
+			name: "panic-json-int2",
+			input: struct {
+				Value int `json:"value" default:"1.23"`
+			}{},
+			panics: `invalid integer tag value '1.23' for field 'Value': schema is invalid`,
+		},
+		{
+			name: "panic-json-array",
+			input: struct {
+				Value []int `json:"value" default:"true"`
+			}{},
+			panics: `invalid array tag value 'true' for field 'Value': schema is invalid`,
+		},
+		{
+			name: "panic-json-array-value",
+			input: struct {
+				Value []string `json:"value" default:"[true]"`
+			}{},
+			panics: `invalid string tag value 'true' for field 'Value[0]': schema is invalid`,
+		},
+		{
+			name: "panic-json-array-value",
+			input: struct {
+				Value []int `json:"value" default:"[true]"`
+			}{},
+			panics: `invalid number tag value 'true' for field 'Value[0]': schema is invalid`,
+		},
+		{
+			name: "panic-json-object",
+			input: struct {
+				Value struct {
+					Foo string `json:"foo"`
+				} `json:"value" default:"true"`
+			}{},
+			panics: `invalid object tag value 'true' for field 'Value': schema is invalid`,
+		},
+		{
+			name: "panic-json-object-field",
+			input: struct {
+				Value struct {
+					Foo string `json:"foo"`
+				} `json:"value" default:"{\"foo\": true}"`
+			}{},
+			panics: `invalid string tag value 'true' for field 'Value.foo': schema is invalid`,
 		},
 	}
 
@@ -543,7 +708,7 @@ func (o OmittableNullable[T]) Schema(r huma.Registry) *huma.Schema {
 
 func TestCustomUnmarshalType(t *testing.T) {
 	type O struct {
-		Field OmittableNullable[int] `json:"field" maximum:"10"`
+		Field OmittableNullable[int] `json:"field" maximum:"10" example:"5"`
 	}
 
 	var o O
@@ -553,6 +718,7 @@ func TestCustomUnmarshalType(t *testing.T) {
 	s := r.Schema(reflect.TypeOf(o), false, "")
 	assert.Equal(t, "integer", s.Properties["field"].Type, s)
 	assert.Equal(t, Ptr(float64(10)), s.Properties["field"].Maximum, s)
+	assert.Equal(t, float64(5), s.Properties["field"].Examples[0], s.Properties["field"])
 
 	// Confirm the field works as expected when loading JSON.
 	o = O{}

--- a/validate_test.go
+++ b/validate_test.go
@@ -960,11 +960,32 @@ var validateTests = []struct {
 		errs:  []string{"expected required property value to be present"},
 	},
 	{
-		name: "enum success",
+		name: "enum string success",
 		typ: reflect.TypeOf(struct {
 			Value string `json:"value" enum:"one,two"`
 		}{}),
 		input: map[string]any{"value": "one"},
+	},
+	{
+		name: "enum int success",
+		typ: reflect.TypeOf(struct {
+			Value int `json:"value" enum:"1,5,9"`
+		}{}),
+		input: map[string]any{"value": 1.0},
+	},
+	{
+		name: "enum uint16 success",
+		typ: reflect.TypeOf(struct {
+			Value uint16 `json:"value" enum:"1,5,9"`
+		}{}),
+		input: map[string]any{"value": 1.0},
+	},
+	{
+		name: "enum array success",
+		typ: reflect.TypeOf(struct {
+			Value []int `json:"value" enum:"1,5,9"`
+		}{}),
+		input: map[string]any{"value": []any{1.0}},
 	},
 	{
 		name: "expected enum",

